### PR TITLE
Track generated OpenAPI definition as JSON in the source tree (#2987)

### PR DESCRIPTION
**Pull request by amarjandu**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: lambdas/service/app.py [hunk 1]
```diff
@@ -471,19 +471,27 @@ def openapi():
 }
 
 fast_health_keys = {
-    **{
-        prop.key: format_description(prop.description)
-        for prop in HealthController.fast_properties['service']
-    },
-    **health_up_key
+    p: d for p, d in sorted(
```
Comment (jessebrennan) [inline lambdas/service/app.py:474]: If changing `HealthController.fast_properties` from `Mapping[str, Set[health_property]]` to `Mapping[str, Tuple[health_property]]` solves the inconsistent ordering, I think that would be a more elegant solution.

File: lambdas/service/app.py [hunk 2]
```diff
@@ -471,19 +471,27 @@ def openapi():
...
-    **{
-        prop.key: format_description(prop.description)
-        for prop in HealthController.all_properties
-    },
-    **health_up_key
+    p: d for p, d in sorted(
+        {
+            **{
+                prop.key: format_description(prop.description)
+                for prop in HealthController.all_properties
```
Comment (jessebrennan) [inline lambdas/service/app.py:490]: For `HealthController.all_properties`, I think the definition could be change to a tuple safely too, and would avoid the need for these changes. If it can't be done safely, then I think it would still make sense to sort at the definition instead of here.

**Review by jessebrennan — CHANGES_REQUESTED** [review]

File: lambdas/service/app.py (hunk 1 above)
Comment (amarjandu) [reply lambdas/service/app.py]: changing from set to a tuple does not resolve the non-deterministic ordering. :|

File: src/azul/health.py [hunk 3]
```diff
@@ -231,16 +231,16 @@ def cached_health(self) -> JSON:
                 body = cache['health']
         return body
 
-    fast_properties: Mapping[str, Set[health_property]] = {
-        'indexer': {
+    fast_properties: Mapping[str, Tuple[health_property]] = {
+        'indexer': (
```
Comment (jessebrennan) [inline src/azul/health.py:235]: Since the reasoning for the changing the properties to tuples isn't obvious, you could consider making it a non-fixup commit that occurs before the rest of the changes.

Giving the new commit a title like "Make ordering of health properties in OpenAPI spec deterministic" would make the reason behind these changes clear.

I think this is optional though.

**Review by jessebrennan — CHANGES_REQUESTED** [review]
These are mostly suggestions, but I think using `del` is a worthwhile change.

File: src/azul/chalice.py [hunk 4]
```diff
@@ -104,14 +106,21 @@ def spec(self) -> JSON:
...
-            'servers': [{'url': self.self_url('/')}]
+            ]
         }
+        if static:
+            servers = []
+            # noinspection PyTypeChecker
+            spec['info'].pop('title')
+        else:
+            servers = [{'url': self.self_url('/')}]
+        spec['servers'] = servers
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/chalice.py:122]: Looks like there is still a type warning with `del`...

Screen Shot 2021-04-26 at 9 35 57 AM

**Comment by jessebrennan** [comment]
Weird, I didn't get that.

File: src/azul/chalice.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/chalice.py]: resolved type warnings with type hint for `spec` during decoration.

[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/chalice.py]: @hannes-ucsc what's your preference about a two inspections to the value of `static` as jesse suggested vs having a larger block of code to manage?

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -54,6 +54,7 @@ test:
   script:
     - make format  # Any ill-formatted sources ...
     - test "$deployment" != sandbox || make requirements_update  # ... or stale transitive dependencies ...
+    - make openapi # ... or uncommit changes to the OpenAPI document
```
Comment (jessebrennan) [inline .gitlab-ci.yml:57]:
```suggestion
    - make openapi # ... or uncommit changes to the OpenAPI document ...
```

File: .travis.yml [hunk 6]
```diff
@@ -53,6 +53,7 @@ jobs:
         # to `docker run` and bind-mount an /etc/passwd that maps that to `developer`.
         # Currently I'm unclear why `make format` works unchanged on Docker
         - sudo chgrp -R 1000 . && make format && sudo chgrp -R $(id -g) .
+        - make openapi
```
Comment (jessebrennan) [inline .travis.yml:56]: This is unnecessary after you rebase.

File: README.md [hunk 7]
```diff
@@ -3,73 +3,6 @@
 The Azul project contains the components that together serve as the backend to
 Boardwalk, a web application for browsing genomic data sets.
 
-# Table of Contents
```
Comment (jessebrennan) [inline README.md:6]: Isn't deleting this covered in another PR or ticket somewhere?

It looks like it is, but it's showing up here because this is chained on that PR? Make sure you set the proper base branch, labels, etc. if I'm correct and this should be a chain.

**Review by jessebrennan — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (amarjandu) [reply README.md:6]: Its a freebie, the commit issue is referenced in the title. The base should be develop. Is there something I messed up? I thought the blocking relation would be appropriate.

Comment (jessebrennan) [reply README.md:6]: Sorry, my mistake. I thought this was already in a separate PR. With the new Zenhub changes you could try linking the freebie as an experiment, but that's not approved process.

**Comment by amarjandu** [comment]
@jessebrennan dude that new CI testing is fast. 🚗

**Review by jessebrennan — APPROVED** [review]
I worry it may accidentally ignore something it's not supposed to, or conflict with a real gitignore file. Consider just removing it, or perhaps change the name so git won't still see it as a `.gitignore` file. 

This looks good though, and I'm approving because I think this is optional.

File: src/azul/health.py [hunk 8]
```diff
@@ -251,9 +251,9 @@ def update_cache(self) -> None:
         self.storage_service.put(object_key=f'health/{self.lambda_name}',
                                  data=json.dumps(health_object).encode())
 
-    all_properties: Set[health_property] = {
+    all_properties: Tuple[health_property] = tuple(
         p for p in locals().values() if isinstance(p, health_property)
```
Comment (hannes-ucsc) [inline src/azul/health.py:255]: This means that the ordering of the result depends on the ordering of the entries in the dictionary returned by locals() which I assume is in the order with which the symbols are defined in the module. While it is deterministic, I think that's a bit arbitrary. I also don't like losing the uniqueness guarantees that come with `Set`. And I don't like how this now mixes `Tuple` for `all_properties` and `Set` for `all_keys`. If we have to use a sequence, which I hope we don't, I would prefer `list` over `tuple`.

TL;DR I think we should address the non-determinism at a higher level by sorting explicitly in app.py.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: lambdas/service/app.py (hunk 2 above)
Comment (hannes-ucsc) [reply lambdas/service/app.py]: https://github.com/DataBiosphere/azul/pull/2997#discussion_r625339700

File: lambdas/service/app.py (hunk 1 above)
Comment (hannes-ucsc) [reply lambdas/service/app.py]: https://github.com/DataBiosphere/azul/pull/2997#discussion_r625339700

**Comment by hannes-ucsc** [comment]
Be sure to run the openapi target with sandbox `dev` and `prod` selected. It shouldn't dirty the file in either case.

File: src/azul/health.py (hunk 8 above)
Comment (hannes-ucsc) [reply src/azul/health.py:255]: Which was your original approach, anyways, I realize now.

Comment (jessebrennan) [reply src/azul/health.py:255]: My bad for the unfavorable suggestion. I'm confused why `list` is preferred to `tuple` if it's not something you plan on mutating, but I suppose that's besides the point.

File: lambdas/service/app.py [hunk 9]
```diff
@@ -416,19 +416,27 @@ def openapi():
 }
 
 fast_health_keys = {
-    **{
-        prop.key: format_description(prop.description)
-        for prop in HealthController.fast_properties['service']
-    },
-    **health_up_key
+    p: d for p, d in sorted(
```
Comment (hannes-ucsc) [inline lambdas/service/app.py:419]: {k:v for k,v in i} == dict(i)

File: lambdas/service/app.py [hunk 10]
```diff
@@ -416,19 +416,27 @@ def openapi():
...
+        {
+            **{
+                prop.key: format_description(prop.description)
+                for prop in HealthController.fast_properties['service']
+            },
+            **health_up_key
+        }.items(),
+        key=lambda key: key[0])
```
Comment (hannes-ucsc) [inline lambdas/service/app.py:427]: You are sorting an iterable of (k,v) tuples, and the keys (k) are unique, so the values (v) are never needed to break a tie. Since tuples are directly comparable, the default key should work. Also, to call the argument to a key function "key" is misleading. And look at `itemgetter` from `operators`. But you won't need it.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: scripts/generate_openapi_document.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    mock,
+)
+
+mock_catalogs = mock.patch.dict(os.environ,
+                                AZUL_CATALOGS='hca:dcp2:repository/tdr:metadata/hca')
+
+
+def main():
+    with mock_catalogs:
+        # imports to config must be done after environment variables are patched
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply scripts/generate_openapi_document.py:14]: And try patching `azul.config` to avoid this.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: scripts/generate_openapi_document.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+            config,
+        )
+        doc_path = os.path.join(config.project_root, 'lambdas/service/openapi.json')
+
+        # add the project root to allow app.py to be located
+        sys.path.append(config.project_root)
+        from lambdas.service.app import (
+            app,
+        )
+        mock_server = mock.patch('lambdas.service.app.app.self_url',
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply scripts/generate_openapi_document.py:25]: https://github.com/DataBiosphere/azul/blob/dbd1ed0653a0a1784328e1ffe4ec5c41df7ffb75/src/azul/chalice.py#L110

but ill rename it.

File: scripts/generate_openapi_document.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+                                 return_value='')
+
+        with mock_server:
+            openpi_spec = app.spec()
+            del openpi_spec['info']['title']
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply scripts/generate_openapi_document.py:30]: It's deployment specific, was having issues with patching the `config.service_name`.

Screen Shot 2021-05-18 at 10 55 05 AM

[the remark this answers is not in the document]
Comment (amarjandu) [reply scripts/generate_openapi_document.py:30]: still having problems with patching the config.service_name, ended up figuring out another way to do this.

File: scripts/generate_openapi_document.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (amarjandu) [reply scripts/generate_openapi_document.py:14]: got around the issue after changing how the service module was being imported.

File: README.md [hunk 14]
```diff
@@ -1844,3 +1777,16 @@ make -C lambdas/service local
 The script serves the Swagger editor locally at a URL where your current version
 of the API documentation is visible. Change the docs in `azul/service/app.py`,
 save, refresh the page, and your changes will appear immediately.
+
+
+## 13.2 Tracking changes to the OpenAPI definition document
```
Comment (hannes-ucsc) [inline README.md:1782]:
```suggestion
## 13.2 Tracking changes to the OpenAPI definition
```

File: README.md [hunk 15]
```diff
@@ -1844,3 +1777,16 @@ make -C lambdas/service local
...
+
+Changes to the OpenAPI document are tracked. When making changes that affect the
```
Comment (hannes-ucsc) [inline README.md:1784]:
```suggestion
Changes to the OpenAPI definition are tracked in the source tree. When making changes that affect the
```

File: README.md [hunk 16]
```diff
@@ -1844,3 +1777,16 @@ make -C lambdas/service local
...
+document, run:
```
Comment (hannes-ucsc) [inline README.md:1785]:
```suggestion
definition, run:
```

File: README.md [hunk 17]
```diff
@@ -1844,3 +1777,16 @@ make -C lambdas/service local
...
+
+```
+make openapi
+```
+
+and commit any modifications to `openapi.json` file. Failure to do so will 
+result in a CI pipeline failure.
```
Comment (hannes-ucsc) [inline README.md:1792]:
```suggestion
and commit any modifications to the `openapi.json` file. Failure to do so will 
break continuous integration during `make check_clean`.
```

File: scripts/generate_openapi_document.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+from azul.modules import (
+    load_app_module,
+)
+
+
+def main():
+    mock_catalogs = mock.patch.dict(os.environ,
+                                    AZUL_CATALOGS='hca:dcp2:repository/tdr:metadata/hca')
+    with mock_catalogs:
+        service = load_app_module('service')
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:26]:
```suggestion
        app_module = load_app_module('service')
```

**Review by hannes-ucsc — COMMENTED** [review]
The mocking looks really complicated. Why can't you just patch `azul.config`?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
https://github.com/DataBiosphere/azul/pull/2997#pullrequestreview-665914681

File: README.md (hunk 14 above)
Comment (amarjandu) [reply README.md:1782]: I'll change this, but i think the "official" term is document. 
https://swagger.io/specification/#definitions

**Comment by amarjandu** [comment]
> The mocking looks really complicated. Why can't you just patch `azul.config`?

I've attempted to mock out the `azul.config` and `app_module` to simplify some of these out, and have not been able to patch targets in a simpler way. I'm going to ask for ~10min of time after parking lot to help simplify the `service_function_name`.

File: scripts/generate_openapi_document.py [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    write_file_atomically,
+)
+from azul.modules import (
+    load_app_module,
+)
+
+
+def main():
+    mock_catalogs = mock.patch.dict(os.environ,
+                                    AZUL_CATALOGS='hca:dcp2:repository/tdr:metadata/hca')
```
Comment (amarjandu) [inline scripts/generate_openapi_document.py:20]: self-note: Watch https://github.com/DataBiosphere/azul/pull/3109

File: lambdas/service/app.py [hunk 20]
```diff
@@ -498,21 +498,25 @@ def openapi():
...
 }
 
-fast_health_keys = {
-    **{
-        prop.key: format_description(prop.description)
-        for prop in HealthController.fast_properties['service']
-    },
-    **health_up_key
-}
+fast_health_keys = dict(sorted(
```
Comment (hannes-ucsc) [inline lambdas/service/app.py:501]: Sorry for waffling on this but I think I prefer using tuples:

```diff
Index: lambdas/service/app.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/service/app.py b/lambdas/service/app.py
--- a/lambdas/service/app.py	(revision 445cd50de9a6b9b3893b0f77583ab7fc7eeae30a)
+++ b/lambdas/service/app.py	(date 1624309970166)
@@ -498,25 +498,21 @@
     '''),
 }
 
-fast_health_keys = dict(sorted(
-    {
-        **{
-            prop.key: format_description(prop.description)
-            for prop in HealthController.fast_properties['service']
-        },
-        **health_up_key
-    }.items()
-))
+fast_health_keys = {
+    **{
+        prop.key: format_description(prop.description)
+        for prop in HealthController.fast_properties['service']
+    },
+    **health_up_key
+}
 
-health_all_keys = dict(sorted(
-    {
-        **{
-            prop.key: format_description(prop.description)
-            for prop in HealthController.all_properties
-        },
-        **health_up_key
-    }.items()
-))
+health_all_keys = {
+    **{
+        prop.key: format_description(prop.description)
+        for prop in HealthController.all_properties
+    },
+    **health_up_key
+}
 
 
 def health_spec(health_keys: dict):
Index: src/azul/health.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/health.py b/src/azul/health.py
--- a/src/azul/health.py	(revision 445cd50de9a6b9b3893b0f77583ab7fc7eeae30a)
+++ b/src/azul/health.py	(date 1624311789063)
@@ -7,6 +7,7 @@
 import json
 import time
 from typing import (
+    AbstractSet,
     Iterable,
     Mapping,
     Set,
@@ -231,16 +232,16 @@
                 body = cache['health']
         return body
 
-    fast_properties: Mapping[str, Set[health_property]] = {
-        'indexer': {
+    fast_properties: Mapping[str, Iterable[health_property]] = {
+        'indexer': (
             elasticsearch,
             queues,
             progress
-        },
-        'service': {
+        ),
+        'service': (
             elasticsearch,
             api_endpoints,
-        }
+        )
     }
 
     def _as_json_fast(self) -> JSON:
@@ -251,10 +252,10 @@
         self.storage_service.put(object_key=f'health/{self.lambda_name}',
                                  data=json.dumps(health_object).encode())
 
-    all_properties: Set[health_property] = {
+    all_properties: Iterable[health_property] = {
         p for p in locals().values() if isinstance(p, health_property)
     }
 
     @classmethod
-    def all_keys(cls) -> Set[str]:
-        return {p.key for p in cls.all_properties}
+    def all_keys(cls) -> AbstractSet[str]:
+        return frozenset(p.key for p in cls.all_properties)
```

File: scripts/generate_openapi_document.py [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    mock_service_endpoint = mock.patch.object(target=config,
+                                              attribute='service_endpoint',
+                                              return_value='localhost')
+    mock_config_service_name = mock.patch.object(target=config,
+                                                 attribute='service_function_name',
+                                                 return_value='azul_service')
+    with mock_catalogs, mock_config_service_name, mock_service_endpoint:
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:27]:
```diff
Index: scripts/generate_openapi_document.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/scripts/generate_openapi_document.py b/scripts/generate_openapi_document.py
--- a/scripts/generate_openapi_document.py	(revision 445cd50de9a6b9b3893b0f77583ab7fc7eeae30a)
+++ b/scripts/generate_openapi_document.py	(date 1624312407376)
@@ -1,7 +1,7 @@
 import json
 import os
-from unittest import (
-    mock,
+from unittest.mock import (
+    patch,
 )
 
 from azul import (
@@ -16,15 +16,16 @@
 
 
 def main():
-    mock_catalogs = mock.patch.dict(os.environ,
-                                    AZUL_CATALOGS='hca:dcp2:repository/tdr:metadata/hca')
-    mock_service_endpoint = mock.patch.object(target=config,
-                                              attribute='service_endpoint',
-                                              return_value='localhost')
-    mock_config_service_name = mock.patch.object(target=config,
-                                                 attribute='service_function_name',
-                                                 return_value='azul_service')
-    with mock_catalogs, mock_config_service_name, mock_service_endpoint:
+    with (
+        patch.dict(os.environ,
+                   AZUL_CATALOGS='hca:dcp2:repository/tdr:metadata/hca'),
+        patch.object(target=config,
+                     attribute='service_function_name',
+                     return_value='azul_service'),
+        patch.object(target=config,
+                     attribute='service_endpoint',
+                     return_value='localhost')
+    ):
         app_module = load_app_module('service')
         doc_path = os.path.join(config.project_root, 'lambdas/service/openapi.json')
         with write_file_atomically(doc_path) as file:
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
No bump.

File: scripts/generate_openapi_document.py (hunk 21 above)
Comment (amarjandu) [reply scripts/generate_openapi_document.py:27]: mhm. cant use a tuple here because of an attribute error....
```
python /Users/amar/dev/hca/azul/scripts/generate_openapi_document.py
Traceback (most recent call last):
  File "/Users/amar/dev/hca/azul/scripts/generate_openapi_document.py", line 36, in <module>
    main()
  File "/Users/amar/dev/hca/azul/scripts/generate_openapi_document.py", line 19, in main
    with (
AttributeError: __enter__
make: *** [openapi] Error 1
```

Comment (amarjandu) [reply scripts/generate_openapi_document.py:27]: @hannes-ucsc tuples within `with` statements wont work until 3.10 lands. https://www.python.org/dev/peps/pep-0617/#migration-plan

File: lambdas/service/app.py (hunk 20 above)
Comment (amarjandu) [reply lambdas/service/app.py:501]: @hannes-ucsc I think there still needs to be sorting done on the keys within `service.app.health_*_keys`...

File: scripts/generate_openapi_document.py (hunk 21 above)
Comment (hannes-ucsc) [reply scripts/generate_openapi_document.py:27]: That link is a bit unspecific. I'm assuming you're referring to this bug: 

https://bugs.python.org/issue12782

Comment (hannes-ucsc) [reply scripts/generate_openapi_document.py:27]: Just nest with statements in that case. The indirection through long variable names is unwieldy.

File: lambdas/service/app.py (hunk 20 above)
Comment (hannes-ucsc) [reply lambdas/service/app.py:501]: Why?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by amarjandu** [comment]
@hannes-ucsc 
> Why? 

We want to remove non-deterministic output within the openapi file to prevent CI failures.
```Python
>>> {x for x in (3,2,1)}
{1, 2, 3}
>>> set((3,2,1))
{1, 2, 3}
>>> frozenset((3,2,1))
frozenset({1, 2, 3})
>>> {x:'bar' for x in (3,2,1)}
{3: 'bar', 2: 'bar', 1: 'bar'}
>>> {x:'bar' for x in frozenset((3,2,1))}
{1: 'bar', 2: 'bar', 3: 'bar'}

>>> {x:'bar' for x in frozenset({x for x in (3,2,1)})}
{1: 'bar', 2: 'bar', 3: 'bar'}
```

For `fast_health_keys` this might not be an issue because the dictionary is being composed from the tuple, but for the `all_health_keys` i don't believe the ordering is guaranteed because the dictionary is composed through the frozenset.

**Comment by amarjandu** [comment]
self note: tested multiple deployments env files for changes.

File: scripts/generate_openapi_document.py [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+from azul.files import (
+    write_file_atomically,
+)
+from azul.modules import (
+    load_app_module,
+)
+
+
+def main():
+    with patch.dict(os.environ,
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:19]: I think I asked this before but don't remember the answer. Why do we patch the environment instead of the config here?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
image

Just delete that file. There's nothing in there worth retaining as an anchor for future discovery of the deleting commit. No bump.

File: scripts/generate_openapi_document.py [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+from azul.modules import (
+    load_app_module,
+)
+
+
+def main():
+    with patch.object(target=type(config),
+                      attribute='catalogs',
+                      new_callable=PropertyMock,
+                      return_value={'dcp2': config.Catalog(name='dcp2',
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:23]: W

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: scripts/generate_openapi_document.py [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+def main():
+    with patch.object(target=type(config),
+                      attribute='catalogs',
+                      new_callable=PropertyMock,
+                      return_value={
+                          'dcp2': config.Catalog(name='dcp2',
+                                                 atlas='hca',
+                                                 plugins=dict(repository='tdr',
+                                                              metadata='hca'))
+                      }):
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:28]: This would probably look better if you extracted the dict literal into a variable.

Comment (hannes-ucsc) [reply scripts/generate_openapi_document.py:28]: Adding the assert from your experiments wouldn't hurt either.

File: scripts/generate_openapi_document.py [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+from azul.files import (
+    write_file_atomically,
+)
+from azul.modules import (
+    load_app_module,
+)
+
+
+def main():
+    with patch.object(target=type(config),
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:20]: Add short comment explaining the need for patching a bit.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
No bump.

File: scripts/generate_openapi_document.py (hunk 24 above)
Comment (amarjandu) [reply scripts/generate_openapi_document.py:28]: @danielsotirhos 👀

File: scripts/generate_openapi_document.py [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+def main():
+    catalogs = {
+        'dcp2': config.Catalog(name='dcp2',
+                               atlas='hca',
+                               plugins=dict(repository='tdr',
+                                            metadata='hca'))
+    }
+
+    # To create a normalized OpenAPI document, we patch deployment
+    # specific variables that will be present within the document.
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:28]:
```suggestion
    # To create a normalized OpenAPI document, we patch any
    # deployment-specific variables that affect the document.
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Not seeing the assert I [asked for](https://github.com/DataBiosphere/azul/pull/2997#discussion_r663144250).

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
https://github.com/DataBiosphere/azul/pull/2997/files#r665714592

File: scripts/generate_openapi_document.py [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+                          attribute='service_function_name',
+                          return_value='azul_service'):
+            with patch.object(target=config,
+                              attribute='service_endpoint',
+                              return_value='localhost'):
+                app_module = load_app_module('service')
+                assert app_module.config.catalogs == catalogs
+                app_spec = app_module.app.spec()
+                assert app_spec['info']['title'] == 'azul_service'
+                assert one(app_spec['servers'])['url'] == '/'
```
Comment (hannes-ucsc) [inline scripts/generate_openapi_document.py:47]: The asserts should be made as soon as possible. The first assert could be done in the body of the first `with`, for example.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
PR title unnecessarily deviated from issue description in insignificant ways. Since you're operator, you can fix the commit title before pushing to sandbox.
